### DISCUSSION TITLE
Credit based interface protocol in SpinalHDL

### DISCUSSION THREAD

**Discussion by wswslzp** [SpinalHDL/SpinalHDL#804] [Ideas]
The valid-ready protocol has been implemented as `Stream` in Spinal lib. It's a widely used handshake protocol for internal data transfer between blocks closed to each other. 

However, in a large system, the credit-based protocol is preferred for data transfer between blocks that are not physically nearby, even in different chiplets. They need long wires to communicate and retiming register (repeater) to meet timing requirements. One advantage of credit-based over valid-ready is that adding repeaters is much easier. 

A typical credit-based interface includes:
* `valid` and `payload`, from master to slave
* `credit`, from slave to master

Master will send a valid payload if it has a credit. After sending a packet, the master consumes a credit. Slave contains a FIFO/buffer and a credit counter to track its credits. If a slave receives a valid payload and stores it in buffer successfully, it will return a credit back to the master. 

Implementation of the credit-based protocol can be a complement to valid-ready `Stream` protocol. The API can be very similar to the existing `Stream`. I think this will improve SpinalHDL's usability in the large chip design.

**Comment by kleinai** [comment]
I say have at it. This shouldn't be too difficult to pull something basic together. Make sure to add some unit tests!

**Reply by Dolu1990** [reply]
> credit, from slave to master

is it a UInt with the value, or a Bool which pulse each time a credit is provided ?

This system also assume that the master also know at elaboration how many credit maximum there is right ?

**Reply by wswslzp** [reply]
`credit` is a bool signal. 

>This system also assume that the master also know at elaboration how many credit maximum there is right ?

Yes, a parameter like `maxCreditNum` should be known for master and slave.

**Reply by Dolu1990** [reply]
Good to me ^^

**Comment by Dolu1990** [comment]
One question is typo, for instance :

```scala
class CreditBus[T <: Data](hardtype : HardType[T]) extends Bundle with IMasterSlave {
  val flow = Flow(hardtype)
  val credit = Bool()
}
```

Any alternative is welcome

**Comment by Readon** [comment]
> The valid-ready protocol has been implemented as `Stream` in Spinal lib. It's a widely used handshake protocol for internal data transfer between blocks closed to each other.
> 
> However, in a large system, the credit-based protocol is preferred for data transfer between blocks that are not physically nearby, even in different chiplets. They need long wires to communicate and retiming register (repeater) to meet timing requirements. One advantage of credit-based over valid-ready is that adding repeaters is much easier.
> 
> A typical credit-based interface includes:
> 
> * `valid` and `payload`, from master to slave
> * `credit`, from slave to master
> 
> Master will send a valid payload if it has a credit. After sending a packet, the master consumes a credit. Slave contains a FIFO/buffer and a credit counter to track its credits. If a slave receives a valid payload and stores it in buffer successfully, it will return a credit back to the master.
> 
> Implementation of the credit-based protocol can be a complement to valid-ready `Stream` protocol. The API can be very similar to the existing `Stream`. I think this will improve SpinalHDL's usability in the large chip design.

Ahh, this is a simplified version of TCP where the credit is updated by ACK packet.

Btw, what is the disadvantage of repeating a Stream by pipeline registers? 
As I know, cascading Stream with skid buffer can extend the distance and relax the timing.

**Reply by Dolu1990** [reply]
> Btw, what is the disadvantage of repeating a Stream by pipeline registers?

Mostly, cutting combinatorial path on the ready signal is "expensive" in FPGA. It cost datawidth x (1FF + 1MUX)

skid buffer =>  stop sending on "almost full" ?

**Reply by wswslzp** [reply]
Yes, retiming on ready path has to include skid buffer, which is expensive also in asic design. More registers and gates will be added when cutting tens of stages. 

In addition, adding skid buffers on Stream is not as easy as credit-based bus and is unfriendly to backend design (especially for asic design).

**Reply by Readon** [reply]
That would be good! This might be another Stream.
